fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -350,17 +355,46 @@ public CompletableFuture<Boolean> markIndexReadable(boolean markReadablePlease)
         if (!markReadablePlease) {
             return AsyncUtil.READY_FALSE; // they didn't say please..
         }
-        AtomicBoolean allReadable = new AtomicBoolean(true);
-        return getRunner().runAsync(context -> openRecordStore(context)
-                .thenCompose(store -> forEachTargetIndex(index ->
-                        store.markIndexReadable(index).thenApply(built -> {
-                            if (!built) {
-                                allReadable.set(false);
-                            }
-                            return null;
-                        }))
-                        .thenApply(ignore -> allReadable.get())
-                ), common.indexLogMessageKeyValues("IndexingBase::markReadable"));
+
+        AtomicReference<RecordIndexUniquenessViolation> uniquenessViolationExRef = new AtomicReference<>();
+        AtomicBoolean anythingChanged = new AtomicBoolean(false);
+
+        // Mark each index readable in its own (retriable, parallel) transaction. If one target fails to become
+        // readable, it should not affect the others.
+        return forEachTargetIndex(index ->
+                markIndexReadableSingleTarget(index, anythingChanged, uniquenessViolationExRef)
+        ).thenApply(ignore -> {
+            RecordIndexUniquenessViolation ex = uniquenessViolationExRef.get();
+            if (ex != null) {
+                throw ex;
+            }
+            return anythingChanged.get();
+        });
+    }
+
+    private CompletableFuture<Boolean> markIndexReadableSingleTarget(Index index, AtomicBoolean anythingChanged,
+                                                                     AtomicReference<RecordIndexUniquenessViolation> uniquenessViolationExRef) {
+        // An extension function to reduce markIndexReadable's complexity
+        return common.getRunner().runAsync(context ->
+                common.getRecordStoreBuilder().copyBuilder().setContext(context).openAsync()
+                        .thenCompose(store ->
+                                policy.shouldAllowUniquePendingState(store) ?
+                                store.markIndexReadableOrUniquePending(index) :
+                                store.markIndexReadable(index))
+        ).handle((changed, ex) -> {
+            if (ex == null) {
+                if (Boolean.TRUE.equals(changed)) {
+                    anythingChanged.set(true);
+                }
+                return changed; // ignored
+            }
+            if (ex instanceof RecordIndexUniquenessViolation) {
```

Comment:
Changed to `RuntimeException` because couldn't make it `Throwable` - if other exception happens, I assume that the next line will throw a cast exception:
```
 runtimeExceptionAtomicReference.set((RuntimeException)ex);
```